### PR TITLE
cacheLifecycle fix for fixedCacheKey in mutationThunk

### LINKED ISSUE #2943: onCacheEntryAdded doesn't work with fixedCacheKey
**onCacheEntryAdded** is not called when **fixedCacheKey** is specified for mutation ([getCacheKey doesn't handle that case](https://github.com/reduxjs/redux-toolkit/blob/f7a828255e0612bed0856a0b6d1b9b8d6d6ac30d/packages/toolkit/src/query/core/buildMiddleware/cacheLifecycle.ts#L254))

```
const [fetchList, list] = rtkApi.useListMutation({
    fixedCacheKey: 'FIXED_CACHE_KEY',
});

...

list: builder.mutation({
  query: () => ({
    method: "POST",
    url: `/some-url`,
  }),
  async onCacheEntryAdded(arg, { cacheDataLoaded }) {
    console.log("log me please");
    await cacheDataLoaded;
    console.log("log again");
  })
})
```

### PR DESCRIPTION
**Pull request by riqts**
fixes #2943 

This PR fixes the initiation of the cacheLifecycle when using fixedCacheKey in mutationThunk and adds a test to verify the creation of a new cache entry when using fixedCacheKey. The getCacheKey function has been refactored to return the fixedCache Key if it exists, otherwise it returns the requestId. A new test has been added to ensure that a new cache entry is created when using fixedCache Key.

### CODE REVIEW

File: packages/toolkit/src/query/tests/useMutation-fixedCacheKey.test.tsx [hunk 1]
```diff
@@ -354,4 +357,34 @@ describe('fixedCacheKey', () => {
...
+      overrideExisting: true,
+      endpoints: (build) => ({
+        send: build.mutation<string, string | Promise<string>>({
+          query: (arg) => arg,
+          onCacheEntryAdded(arg, {}) {
+            onNewCacheEntry(arg)
+          },
+        }),
+      }),
+    })
```
Comment (EskiMojo14) [inline packages/toolkit/src/query/tests/useMutation-fixedCacheKey.test.tsx:372]:
```suggestion
    api.enhanceEndpoints({
      endpoints: {
        send: {
          onCacheEntryEnded: (arg) => onNewCacheEntry(arg)
        }
      }
    })
```

Comment (EskiMojo14) [reply packages/toolkit/src/query/tests/useMutation-fixedCacheKey.test.tsx]: sorry, my suggestion had a typo - edited

the problem with what you have now is that your mock isn't being reset between tests

image

Comment (riqts) [reply packages/toolkit/src/query/tests/useMutation-fixedCacheKey.test.tsx]: All good, I thought the ended was a typo.

Seeing as you're active, I'll ask what the preference here is so I can know for future as I am not an expert on unit testing practices really. Is it preferred that I create the endpoint override within the scope of that single test, or have the functionality added to the initial declaration and add and `afterEach () => {clear mock function}`

Comment (EskiMojo14) [reply packages/toolkit/src/query/tests/useMutation-fixedCacheKey.test.tsx]: i think you should create the mock and call enhanceEndpoints within the single test, and at the end of the test block do 
```ts
    api.enhanceEndpoints({
      endpoints: {
        send: {
          onCacheEntryAdded: undefined,
        },
      },
    })
```
to make sure it doesn't continue onto other tests

Comment (riqts) [reply packages/toolkit/src/query/tests/useMutation-fixedCacheKey.test.tsx]: That's why you get paid the big bucks, Ben. Recent commit should reflect this change.

**Comment by EskiMojo14** [comment]
thanks!

**Pull request merged**